app/channel.go

```diff
@@ -2912,7 +2912,7 @@ func (a *App) MarkChannelsAsViewed(channelIDs []string, userID string, currentSe
 		a.clearPushNotification(currentSessionId, userID, channelID, "")
 	}
 
-	if !collapsedThreadsSupported || !a.IsCRTEnabledForUser(userID) {
+	if *a.Config().ServiceSettings.ThreadAutoFollow && (!collapsedThreadsSupported || !a.IsCRTEnabledForUser(userID)) {
 		if err := a.Srv().Store.Thread().MarkAllAsReadInChannels(userID, channelIDs); err != nil {
 			return nil, model.NewAppError("MarkChannelsAsViewed", "app.channel.update_last_viewed_at.app_error", nil, err.Error(), http.StatusInternalServerError)
```

Comment:
@Willyfrog @koox00 

Here we mark every thread in a channel as read if user doesn't have CRT enabled (or client is not supported). 

My suggestion is that we don't do this if `ThreadAutoFollow` is false. 
A long time ago `ThreadAutoFollow` only decided whether a thread should be auto-followed for a user. Now we're using `ThreadAutoFollow` as the check for whether the server should do the backend book-keeping. 

Therefore, if `ThreadAutoFollow` is false, I think we shouldn't update a user's threads. 

Pros:
1. Better performance. We wont be running a heavy query on every channel view (i.e. every channel switch)

Cons:
1. The user's threads will not be marked as read on viewing a channel. This might lead to a huge number of unread threads when the server finally turns on CRT. I don't think this is huge problem, given that backend bookkeeping is expected not to happen when `ThreadAutoFollow` is false.

Reply:
I agree we are treating `ThreadAutoFollow` as the bookeeping flag, and if so it makes sense the proposal.

Regarding the cons: we should prepare a migration/job when we approach GA for easing the change on big installs.

Reply:
> The user's threads will not be marked as read on viewing a channel. This might lead to a huge number of unread threads when the server finally turns on CRT. I don't think this is huge problem, given that backend bookkeeping is expected not to happen when ThreadAutoFollow is false.

@ashishbhate @Willyfrog if ThreadAutoFollow is off are we even populating the threadmemberships and related tables? I thought the ThreadAutoFollow was the switch that turns off all CRT related backend work - and if that's the case then shouldn't disabling it mean there's no threads to mark as read anyway?

Reply:
@esethna No new CRT threads will be created so the the number of unread threads will not increase. But all older unread threads will remain unread.